### ISSUE TITLE
Animating Rigid Body Motion - What is a good practise?

### ISSUE THREAD

**Issue by SimiPixel** [vispy/vispy#2448]
Dear authors,

i am currently trying to animate a rigid body / kinematic tree. Essentially, a bunch of geometries whose 4x4 Transform i update at every frame to create an animation. Unfortunately, performance suffers. 

Q: What is the best approach here? Is there an established way of accomplishing this?

MWE (it uses jax)
```python
from dataclasses import dataclass, field

import jax
import jax.numpy as jnp
import numpy as np
from jax import random as jrand
from vispy import app, scene
from vispy.scene import MatrixTransform


@dataclass
class Geometry:
    mass: jax.Array
    CoM: jax.Array


@dataclass
class Box(Geometry):
    dim_x: jax.Array
    dim_y: jax.Array
    dim_z: jax.Array
    vispy_kwargs: dict = field(default_factory=lambda: {})


def quat_to_3x3(q: jnp.ndarray) -> jnp.ndarray:
    """Converts quaternion to 3x3 rotation matrix."""
    d = jnp.dot(q, q)
    w, x, y, z = q
    s = 2 / d
    xs, ys, zs = x * s, y * s, z * s
    wx, wy, wz = w * xs, w * ys, w * zs
    xx, xy, xz = x * xs, x * ys, x * zs
    yy, yz, zz = y * ys, y * zs, z * zs

    return jnp.array(
        [
            jnp.array([1 - (yy + zz), xy - wz, xz + wy]),
            jnp.array([xy + wz, 1 - (xx + zz), yz - wx]),
            jnp.array([xz - wy, yz + wx, 1 - (xx + yy)]),
        ]
    )


def quat_random(key: jrand.PRNGKey, batch_shape: tuple[int] = ()) -> jax.Array:
    """Provides a random quaternion, sampled uniformly"""
    shape = batch_shape + (4,)
    qs = jrand.normal(key, shape)
    qs_norm = jnp.linalg.norm(qs, axis=-1)
    return qs / qs_norm[..., None]


@jax.jit
def _transform_4x4(pos, rot, com):
    E = quat_to_3x3(rot)
    M = jnp.eye(4)
    M = M.at[:3, :3].set(E)
    pos = pos + E.T @ com
    T = jnp.eye(4)
    T = T.at[3, :3].set(pos)
    return M @ T


def transform_4x4(pos, rot, com=jnp.zeros((3,))):
    return MatrixTransform(np.asarray(_transform_4x4(pos, rot, com)))


def convert_timestep(timestep):
    return {0.01: 0.02}[timestep]


class VisPyCanvas:
    def __init__(self, show_cs=True, **kwargs):

        self.canvas = scene.SceneCanvas(
            keys="interactive", size=(800, 600), show=True, **kwargs
        )
        self.view = self.canvas.central_widget.add_view()
        self.view.camera = "arcball"
        self.show_cs = show_cs

    def create_visual_element(self, geom: Geometry, **kwargs):
        if isinstance(geom, Box):
            return scene.visuals.Box(
                geom.dim_x, geom.dim_z, geom.dim_y, parent=self.view.scene, **kwargs
            )
        raise NotImplementedError()

    def populate_scene(self, geom_coll: list[list[Geometry]]):
        self.visuals = []
        self.geom_coll = geom_coll
        self._geoms_cs = []

        if self.show_cs:
            scene.visuals.XYZAxis(parent=self.view.scene)

        for geoms_per_link in geom_coll:

            if self.show_cs:
                self._geoms_cs.append(scene.visuals.XYZAxis(parent=self.view.scene))

            visuals_per_link = []
            for geom in geoms_per_link:
                visuals_per_link.append(
                    self.create_visual_element(geom, **geom.vispy_kwargs)
                )
            self.visuals.append(visuals_per_link)

    def toggle_pause(self):
        self.pause = not self.pause

    def reset(self):
        self.pause = True
        self.time = 0
        self.t = 0
        self._update_scene()

    def add_motion_data(
        self, timestep: float, data_pos: jax.Array, data_rot: jax.Array
    ):
        self.data_pos = data_pos
        self.data_rot = data_rot

        self.timer_interval = convert_timestep(timestep)
        self.n_skips = int(self.timer_interval / timestep)
        self.timestep = timestep

        # initializes scence
        self.reset()

    def _get_link_data(self, link_idx: int):

        rot = self.data_rot[self.t][link_idx]
        pos = self.data_pos[self.t][link_idx]
        return rot, pos

    def _get_transform(self, link_idx, geom_idx=None):

        rot, pos = self._get_link_data(link_idx)

        if geom_idx is not None:
            return transform_4x4(pos, rot, self.geom_coll[link_idx][geom_idx].CoM)
        else:
            return transform_4x4(pos, rot)

    def _update_scene_per_link(self, link_idx):
        if self.show_cs:
            transform = self._get_transform(link_idx)
            self._geoms_cs[link_idx].transform = transform

        for geom_idx in range(len(self.visuals[link_idx])):
            transform = self._get_transform(link_idx, geom_idx)
            self.visuals[link_idx][geom_idx].transform = transform

    def _update_scene(self):
        for link_idx in range(len(self.visuals)):
            self._update_scene_per_link(link_idx)

    def on_timer(self, event):
        if self.t > len(self.data_pos):
            self.pause = True

        if self.pause:
            return

        self._update_scene()

        print("time: ", round(self.time, 2))

        self.t += self.n_skips
        self.time += self.timer_interval

    def run(self):
        self.timer = app.Timer(
            self.timer_interval,
            connect=self.on_timer,
            start=True,
        )

        app.run()


if __name__ == "__main__":
    box = Box(
        10.0,
        jnp.array([0.5, 0, 0]),
        1.0,
        0.25,
        0.2,
        vispy_kwargs=dict(edge_color="black", color="white"),
    )

    # generally speaking these geometries would be different
    geom_coll = [[box], [], [box, box], [box]]

    # we have four links
    n_links = len(geom_coll)

    T = 10.0
    timestep = 0.01
    N = int(T / timestep)

    data_pos = jrand.normal(jrand.PRNGKey(2), (N, n_links, 3))
    data_rot = quat_random(jrand.PRNGKey(1), (N, n_links))

    vis = VisPyCanvas()
    vis.populate_scene(geom_coll)
    vis.add_motion_data(timestep, data_pos, data_rot)

    # let's go :)
    vis.toggle_pause()
    vis.run()

    # don't look at the visualization; it might hurt
    # look at the time in the terminal; ideally this should be real-time
    # and we are only doing 50 Hz..

```

**Comment by djhoese** [comment]
Very interesting. I'll start by pointing out the issue you linked from in the `google/brax` repository: It sounds like what you are hoping for in the long run is a visualization that uses data from the TPU/GPU. Currently VisPy is not able to visualize data *directly* from another GPU-based context. For example, using data in a CUDA/cupy calculation in the GPU directly as a texture or buffer in VisPy's OpenGL-based computation. There are issues in this repository to discuss solutions towards getting this to work but no one has had the time or experience to get it to work completely. So we can make your example code faster and faster, but it will still require the data to be copied from GPU (Jax) -> CPU -> GPU (OpenGL). Keep in mind I know almost nothing about Jax.

As for your example: In general you'll want to reduce the number of Visuals you create and work with. See https://vispy.org/faq.html#why-is-my-visualization-slower-when-i-add-more-visual-objects for more information. The Box you are using is actually a CompoundVisual so it actually consists of two Visuals (one MeshVisual for the sides, one for the border). As a Mesh, you could theoretically combine all your boxes into a single Mesh (or two, one for the body of the box, one for the border if you need a border). You might be able to do a similar optimization with the XYZAxis objects as a series of Line visuals, but this might be harder to organize...but 2 Visuals (one Mesh for all boxes, one Line for all XYZAxis) is much better than 100s.

Next optimization: Avoid setting `.transform` over and over again. This sets a new Transform object on the Visual and requires VisPy to recompile the OpenGL shader code underneath. Compiling takes time and requires reuploading the shader code to the GPU. The way around this is to update the properties of the Transform, not the entire transform itself. In your case with a `MatrixTransform` you can do `visual.transform.matrix = new_matrix` and VisPy will only send the new matrix to the GPU, not all of the re-compiled shader code.

After this, we start talking about optimizing things as numpy (jax?) arrays and vectorizing where ever possible so we're not doing python for loops.

**Comment by SimiPixel** [comment]
Thanks for the comment.

So, i (of course) started with the easiest optimization first which is this change `visual.transform.matrix = new_matrix`.

I also added a simple timer that computes the ratio of animation to real time.

- Before: At 50 Hz: Ratio: 0.3
- After: At 50 Hz: Ratio: 0.98
- After: At 100 Hz: Ratio: 0.98

This would work but unfortunately nothing happens. The Geometries do not seem to get updated.

The After MWE:
```python
import time
from dataclasses import dataclass, field

import jax
import jax.numpy as jnp
import numpy as np
from jax import random as jrand
from vispy import app, scene


@dataclass
class Geometry:
    mass: jax.Array
    CoM: jax.Array


@dataclass
class Box(Geometry):
    dim_x: jax.Array
    dim_y: jax.Array
    dim_z: jax.Array
    vispy_kwargs: dict = field(default_factory=lambda: {})


def quat_to_3x3(q: jnp.ndarray) -> jnp.ndarray:
    """Converts quaternion to 3x3 rotation matrix."""
    d = jnp.dot(q, q)
    w, x, y, z = q
    s = 2 / d
    xs, ys, zs = x * s, y * s, z * s
    wx, wy, wz = w * xs, w * ys, w * zs
    xx, xy, xz = x * xs, x * ys, x * zs
    yy, yz, zz = y * ys, y * zs, z * zs

    return jnp.array(
        [
            jnp.array([1 - (yy + zz), xy - wz, xz + wy]),
            jnp.array([xy + wz, 1 - (xx + zz), yz - wx]),
            jnp.array([xz - wy, yz + wx, 1 - (xx + yy)]),
        ]
    )


def quat_random(key: jrand.PRNGKey, batch_shape: tuple[int] = ()) -> jax.Array:
    """Provides a random quaternion, sampled uniformly"""
    shape = batch_shape + (4,)
    qs = jrand.normal(key, shape)
    qs_norm = jnp.linalg.norm(qs, axis=-1)
    return qs / qs_norm[..., None]


@jax.jit
def _transform_4x4(pos, rot, com):
    E = quat_to_3x3(rot)
    M = jnp.eye(4)
    M = M.at[:3, :3].set(E)
    pos = pos + E.T @ com
    T = jnp.eye(4)
    T = T.at[3, :3].set(pos)
    return M @ T


def transform_4x4(pos, rot, com=jnp.zeros((3,))):
    return np.asarray(_transform_4x4(pos, rot, com))


def convert_timestep(timestep):
    return {0.01: 0.01}[timestep]


class VisPyCanvas:
    def __init__(self, show_cs=True, **kwargs):

        self.canvas = scene.SceneCanvas(
            keys="interactive", size=(800, 600), show=True, **kwargs
        )
        self.view = self.canvas.central_widget.add_view()
        self.view.camera = "arcball"
        self.show_cs = show_cs

    def create_visual_element(self, geom: Geometry, **kwargs):
        if isinstance(geom, Box):
            return scene.visuals.Box(
                geom.dim_x, geom.dim_z, geom.dim_y, parent=self.view.scene, **kwargs
            )
        raise NotImplementedError()

    def populate_scene(self, geom_coll: list[list[Geometry]]):
        self.visuals = []
        self.geom_coll = geom_coll
        self._geoms_cs = []

        if self.show_cs:
            scene.visuals.XYZAxis(parent=self.view.scene)

        for geoms_per_link in geom_coll:

            if self.show_cs:
                self._geoms_cs.append(scene.visuals.XYZAxis(parent=self.view.scene))

            visuals_per_link = []
            for geom in geoms_per_link:
                visuals_per_link.append(
                    self.create_visual_element(geom, **geom.vispy_kwargs)
                )
            self.visuals.append(visuals_per_link)

    def toggle_pause(self):
        self.pause = not self.pause

    def reset(self):
        self.pause = True
        self.time = 0
        self.t = 0
        self._update_scene()

    def add_motion_data(
        self, timestep: float, data_pos: jax.Array, data_rot: jax.Array
    ):
        self.data_pos = data_pos
        self.data_rot = data_rot

        self.timer_interval = convert_timestep(timestep)
        self.n_skips = int(self.timer_interval / timestep)
        self.timestep = timestep

        # initializes scence
        self.reset()

    def _get_link_data(self, link_idx: int):

        rot = self.data_rot[self.t][link_idx]
        pos = self.data_pos[self.t][link_idx]
        return rot, pos

    def _get_transform_matrix(self, link_idx, geom_idx=None):

        rot, pos = self._get_link_data(link_idx)

        if geom_idx is not None:
            return transform_4x4(pos, rot, self.geom_coll[link_idx][geom_idx].CoM)
        else:
            return transform_4x4(pos, rot)

    def _update_scene_per_link(self, link_idx):
        if self.show_cs:
            transform_matrix = self._get_transform_matrix(link_idx)
            self._geoms_cs[link_idx].transform.matrix = transform_matrix

        for geom_idx in range(len(self.visuals[link_idx])):
            transform_matrix = self._get_transform_matrix(link_idx, geom_idx)
            self.visuals[link_idx][geom_idx].transform.matrix = transform_matrix

    def _update_scene(self):
        for link_idx in range(len(self.visuals)):
            self._update_scene_per_link(link_idx)

    def on_timer(self, event):
        if self.t > len(self.data_pos):
            self.pause = True

        if self.pause:
            return

        self._update_scene()

        print("time: ", round(self.time, 2))

        self.t += self.n_skips
        self.time += self.timer_interval
        self.realtime = time.time()

    def run(self):
        self.timer = app.Timer(
            self.timer_interval,
            connect=self.on_timer,
            start=True,
        )

        app.run()


if __name__ == "__main__":
    box = Box(
        10.0,
        jnp.array([0.5, 0, 0]),
        1.0,
        0.25,
        0.2,
        vispy_kwargs=dict(edge_color="black", color="white"),
    )

    # generally speaking these geometries would be different
    geom_coll = [[box], [], [box, box], [box]]

    # we have four links
    n_links = len(geom_coll)

    T = 5.0
    timestep = 0.01
    N = int(T / timestep)

    data_pos = jrand.normal(jrand.PRNGKey(2), (N, n_links, 3))
    data_rot = quat_random(jrand.PRNGKey(1), (N, n_links))

    vis = VisPyCanvas()
    vis.populate_scene(geom_coll)
    vis.add_motion_data(timestep, data_pos, data_rot)

    # let's go :)
    vis.toggle_pause()
    start_time = time.time()
    vis.run()
    anim_runtime = vis.realtime - start_time

    print(
        f"It took {anim_runtime}s to animate {vis.time}s.\n\
        The ratio of animation time to realtime is {vis.time / anim_runtime}."
    )

    # don't look at the visualization; it might hurt
    # look at the time in the terminal; ideally this should be real-time
    # and we are only doing 50 Hz..
```

**Comment by djhoese** [comment]
```
The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    cudatoolkit-11.8.0         |      h37601d7_11       635.9 MB  conda-forge
    cudnn-8.4.1.50             |       hed8a83a_0       618.4 MB  conda-forge
    jax-0.4.2                  |     pyhd8ed1ab_0         894 KB  conda-forge
    jaxlib-0.4.1               |cuda112py310hf0bc174_201        41.8 MB  conda-forge
    libabseil-20220623.0       | cxx17_h05df665_6         1.1 MB  conda-forge
    libgrpc-1.51.1             |       h30feacc_0         4.9 MB  conda-forge
    libprotobuf-3.21.12        |       h3eb15da_0         2.1 MB  conda-forge
    nccl-2.14.3.1              |       h0800d71_0       145.2 MB  conda-forge
    opt_einsum-3.3.0           |     pyhd8ed1ab_1          53 KB  conda-forge
    re2-2022.06.01             |       h27087fc_1         191 KB  conda-forge
    scipy-1.10.0               |  py310h8deb116_0        23.4 MB  conda-forge
```

1.5GB to install Jax :fearful: 

Anyway...

I can't seem to get your first example to run. Looks like my installation is bad (on Ubuntu):

```
$ python vispy_support/issue_2448.py
2023-02-01 07:40:56.960807: W external/org_tensorflow/tensorflow/compiler/xla/stream_executor/gpu/asm_compiler.cc:109] Couldn't get ptxas version : FAILED_PRECONDITION: Couldn't get ptxas version string: INTERNAL: Couldn't invoke ptxas --version
2023-02-01 07:40:56.961434: F external/org_tensorflow/tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc:451] ptxas returned an error during compilation of ptx to sass: 'INTERNAL: Failed to launch ptxas'  If the error message indicates that a file could not be written, please verify that sufficient filesystem space is provided.
Aborted (core dumped)
```

It is possible the transform update isn't triggering the update of the visual. Can you do a `canvas.update()` (or `self.update()` maybe?) after all the updates are done and see if that helps?

**Comment by SimiPixel** [comment]
Yeah sorry about that. I should point out though that i am only working on my laptop (no GPU, only Tigerlake iGPU). As such it would be enough to go with <pip install jax jaxlib> which should install cpu-only which does not require all the cuda stuff.

Also, i already found the problem regarding this optimization. I forgot to initialize at least once with `MatrixTransform` (and mutate afterwards).

That's already now functional for my use-case. I might try to optimize some more down the line. 

For reference this is my current snippet
```python
import time
from dataclasses import dataclass, field

import jax
import jax.numpy as jnp
import numpy as np
from jax import random as jrand
from vispy import app, scene
from vispy.scene import MatrixTransform


@dataclass
class Geometry:
    mass: jax.Array
    CoM: jax.Array


@dataclass
class Box(Geometry):
    dim_x: jax.Array
    dim_y: jax.Array
    dim_z: jax.Array
    vispy_kwargs: dict = field(default_factory=lambda: {})


def quat_to_3x3(q: jnp.ndarray) -> jnp.ndarray:
    """Converts quaternion to 3x3 rotation matrix."""
    d = jnp.dot(q, q)
    w, x, y, z = q
    s = 2 / d
    xs, ys, zs = x * s, y * s, z * s
    wx, wy, wz = w * xs, w * ys, w * zs
    xx, xy, xz = x * xs, x * ys, x * zs
    yy, yz, zz = y * ys, y * zs, z * zs

    return jnp.array(
        [
            jnp.array([1 - (yy + zz), xy - wz, xz + wy]),
            jnp.array([xy + wz, 1 - (xx + zz), yz - wx]),
            jnp.array([xz - wy, yz + wx, 1 - (xx + yy)]),
        ]
    )


def quat_random(key: jrand.PRNGKey, batch_shape: tuple[int] = ()) -> jax.Array:
    """Provides a random quaternion, sampled uniformly"""
    shape = batch_shape + (4,)
    qs = jrand.normal(key, shape)
    qs_norm = jnp.linalg.norm(qs, axis=-1)
    return qs / qs_norm[..., None]


@jax.jit
def _transform_4x4(pos, rot, com):
    E = quat_to_3x3(rot)
    M = jnp.eye(4)
    M = M.at[:3, :3].set(E)
    pos = pos + E.T @ com
    T = jnp.eye(4)
    T = T.at[3, :3].set(pos)
    return M @ T


def transform_4x4(pos, rot, com=jnp.zeros((3,))):
    return np.asarray(_transform_4x4(pos, rot, com))


def convert_timestep(timestep):
    Hz = 50
    return {0.01: 1 / Hz}[timestep]


class VisPyCanvas:
    def __init__(self, show_cs=True, **kwargs):

        self.canvas = scene.SceneCanvas(
            keys="interactive", size=(800, 600), show=True, **kwargs
        )
        self.view = self.canvas.central_widget.add_view()
        self.view.camera = "arcball"
        self.show_cs = show_cs

    def create_visual_element(self, geom: Geometry, **kwargs):
        if isinstance(geom, Box):
            return scene.visuals.Box(
                geom.dim_x, geom.dim_z, geom.dim_y, parent=self.view.scene, **kwargs
            )
        raise NotImplementedError()

    def populate_scene(self, geom_coll: list[list[Geometry]]):
        self.visuals = []
        self.geom_coll = geom_coll
        self._geoms_cs = []

        if self.show_cs:
            scene.visuals.XYZAxis(parent=self.view.scene)

        for geoms_per_link in geom_coll:

            if self.show_cs:
                self._geoms_cs.append(scene.visuals.XYZAxis(parent=self.view.scene))

            visuals_per_link = []
            for geom in geoms_per_link:
                visuals_per_link.append(
                    self.create_visual_element(geom, **geom.vispy_kwargs)
                )
            self.visuals.append(visuals_per_link)

    def toggle_pause(self):
        self.pause = not self.pause

    def reset(self):
        self.pause = True
        self.time = 0
        self.t = 0
        self._update_scene(mutate=False)

    def add_motion_data(
        self, timestep: float, data_pos: jax.Array, data_rot: jax.Array
    ):
        self.data_pos = data_pos
        self.data_rot = data_rot

        self.timer_interval = convert_timestep(timestep)
        self.n_skips = int(self.timer_interval / timestep)
        self.timestep = timestep

        # initializes scence
        self.reset()

    def _get_link_data(self, link_idx: int):

        rot = self.data_rot[self.t][link_idx]
        pos = self.data_pos[self.t][link_idx]
        return rot, pos

    def _get_transform_matrix(self, link_idx, geom_idx=None):

        rot, pos = self._get_link_data(link_idx)

        if geom_idx is not None:
            return transform_4x4(pos, rot, self.geom_coll[link_idx][geom_idx].CoM)
        else:
            return transform_4x4(pos, rot)

    def _update_scene_per_link(self, link_idx, mutate: bool):
        if self.show_cs:
            transform_matrix = self._get_transform_matrix(link_idx)
            cs = self._geoms_cs[link_idx]
            if not mutate:
                cs.transform = MatrixTransform(transform_matrix)
            else:
                cs.transform.matrix = transform_matrix

        for geom_idx in range(len(self.visuals[link_idx])):
            transform_matrix = self._get_transform_matrix(link_idx, geom_idx)
            visual = self.visuals[link_idx][geom_idx]
            if not mutate:
                visual.transform = MatrixTransform(transform_matrix)
            else:
                visual.transform.matrix = transform_matrix

    def _update_scene(self, mutate: bool = True):
        for link_idx in range(len(self.visuals)):
            self._update_scene_per_link(link_idx, mutate)

    def on_timer(self, event):
        if self.t > len(self.data_pos):
            self.pause = True

        if self.pause:
            return

        self._update_scene()

        print("time: ", round(self.time, 3))

        self.t += self.n_skips
        self.time += self.timer_interval
        self.realtime = time.time()

    def run(self):
        self.timer = app.Timer(
            self.timer_interval,
            connect=self.on_timer,
            start=True,
        )

        app.run()


if __name__ == "__main__":
    box = Box(
        10.0,
        jnp.array([0.5, 0, 0]),
        1.0,
        0.25,
        0.2,
        vispy_kwargs=dict(edge_color="black", color="white"),
    )

    # generally speaking these geometries would be different
    geom_coll = [[box], [], [box, box], [box]]

    # we have four links
    n_links = len(geom_coll)

    T = 5.0
    timestep = 0.01
    N = int(T / timestep)

    data_pos = jrand.normal(jrand.PRNGKey(2), (N, n_links, 3))
    data_rot = quat_random(jrand.PRNGKey(1), (N, n_links))

    vis = VisPyCanvas()
    vis.populate_scene(geom_coll)
    vis.add_motion_data(timestep, data_pos, data_rot)

    # let's go :)
    vis.toggle_pause()
    start_time = time.time()
    vis.run()
    anim_runtime = vis.realtime - start_time

    print(
        f"It took {anim_runtime}s to animate {vis.time}s.\n\
        The ratio of animation time to realtime is {vis.time / anim_runtime}."
    )

    # don't look at the visualization; it might hurt
    # look at the time in the terminal; ideally this should be real time
    # for few boxes it now is (almost) real time
```

**Comment by djhoese** [comment]
Just for my own sanity and since you said "down the line", I'm going to close this for now. I think we've got some ideas in this thread for people looking for similar solutions. Feel free to continue commenting after it is closed and if we feel we need to reopen this we can (but a new issue might be better).
